api/v1alpha1/awsloadbalancercontroller_types.go

```diff
@@ -88,6 +88,15 @@ type AWSLoadBalancerControllerSpec struct {
 	// +optional
 	EnabledAddons []AWSAddon `json:"enabledAddons,omitempty"` // indicates which AWS addons should be disabled.
 
+	// Credentials is a reference to a secret containing
+	// the AWS credentials to be used by the controller.
+	// If this field is empty - the credentials will be
+	// requested using the Cloud Credentials API,
+	// see https://docs.openshift.com/container-platform/4.11/authentication/managing_cloud_provider_credentials/about-cloud-credential-operator.html.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Credentials *SecretReference `json:"credentials,omitempty"`
```

Comment:
No, I don't think being non Openshift specific was a goal. I was thinking more of a) future enhancements which would mean changing the type in case of the usage of `github.com/openshift/api/config/v1#SecretNameReference`, b) being self sufficient: you import the api package and all is inside, no additional packages are needed.

> Does this field need to be a pointer type?

Not necessary. My logic was: since it's optional it would very convenient to check for its existence by just comparing to `nil`.

Reply:
> > Does this field need to be a pointer type?
> 
> Not necessary. My logic was: since it's optional it would very convenient to check for its existence by just comparing to `nil`.

It's about as easy to check `len(credentials.Name) == 0` as to check `credentials == nil`.  [The guidance in the OpenShift API Conventions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#optional-fields-should-not-be-pointers-in-configuration-apis) is to avoid using pointers when they aren't necessary.  

Granted, this is an addon operator, and maybe that weakens the OpenShift API Conventions' applicability here.  Is there a convention for addon operators that takes precedence here?

Reply:
Agree that checking for `len(credentials.Name) == 0` is as simple as checking for `nil`. Also, I'm not aware of any convention for the addon operators. Btw ALBO targets OpenShift only, so I'd say that the OpenShift API conventions are quite applicable here.

However, I don't think I agree with this particular recommendation (neither do the KubeAPI Conventions). It would make sense for a required field but for an optional field it's still quite useful to know whether the field was set by the user or not. It may be not important to distinguish the zero from the nil value now when only `name` exists in `SecretReference`, but I can imagine quite valid cases when, for instance, another field is added to `SecretReference` and `name` field would need to be defaulted. The non pointer credentials field would complicate things then.

Also, my choice was slightly biased by the existing fields of the API: `Config *AWSLoadBalancerDeploymentConfig`, also optional field.

Reply:
> However, I don't think I agree with this particular recommendation (neither do the KubeAPI Conventions). 

Right, it's explicitly called out in the OpenShift API Conventions because it deviates from the Kubernetes API Conventions, which we otherwise follow.  Moreover, the documentation for the OpenShift conventions explains our justification for preferring non-pointer types.  

> It would make sense for a required field but for an optional field it's still quite useful to know whether the field was set by the user or not. It may be not important to distinguish the zero from the nil value now when only `name` exists in `SecretReference`, but I can imagine quite valid cases when, for instance, another field is added to `SecretReference` and `name` field would need to be defaulted.

It's hard for me to imagine what field we might add (maybe namespace?) or why we would need to default `name` in that situation.  

I won't insist on changing the type to a non-pointer type (although maybe the API review team would); I'm just trying to explain the OpenShift conventions and understand your argument.